compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -421,6 +421,16 @@ fn instantiate_and_check_impossible_predicates<'tcx>(
     result
 }
 
+/// Checks whether an item is impossible to reference.
+#[instrument(level = "debug", skip(tcx), ret)]
+fn is_impossible_item<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
+    let param_env = tcx.param_env_reveal_all_normalized(def_id);
```

Comment:
This is kind of strange. Can you just use `elaborate(tcx.predicates_of().instantiate_identity).collect()`? It shouldn't matter to eagerly fold opaques -- we're already processing `impossible_predicates` with a `reveal_all` param-env.

Reply:
That doesn't work, because it would filter out local predicates like `T::Assoc: Sized`, even if they normalize to global predicates like `[u8]: Sized`, which is exactly the case here:

```rust
impl<T> Adapter for T
{
    type A = ApiS;
    fn open() -> OpenDevice<Self::A>
    where
        <Self::A as Api>::Device: Sized,
      // ^^^^ This `Self` is an alias for the type parameter `T`
      // The `predicates_of` has the "local" predicate `<T::A as Api>::Device: Sized`,
      // but the `param_env` (with or without reveal all) has the "global" predicate `[u8]: Sized`
    {
        unreachable!()
    }
}
```

So, we need to normalize the `predicates_of` before filtering out `!clause.has_param()`. And the the normalized + elaborated `predicates_of` is exactly the `param_env`. I'm using `param_env_reveal_all_normalized` instead of `param_env` here as an optimization, exactly because `impossible_predicates` will normalize in the empty reveal-all env anyway.

Reply:
Okay. Still feels a bit hacky that you're using `param_env.caller_bounds()`. This should at least be documented.

> I'm using `param_env_reveal_all_normalized` instead of `param_env` here as an optimization, exactly because `impossible_predicates` will normalize in the empty reveal-all env anyway.

I don't think this is a particularly compelling optimization, especially because the `param_env` query has definitely been fetched for all items we care about, and the likelihood of having an opaque in a where clause anyways is very small. I'd prefer if we write the code more explicitly unless it's proven to be hot, which I doubt is the case for this code, since it's gonna be only running once per item.

Reply:
> Still feels a bit hacky that you're using `param_env.caller_bounds()` as a substitute for, e.g. `normalize_erasing_regions`-ing the elaborated caller bounds or something.

Do you mean `normalize_erasing_regions`-ing the elaborated `predicates_of`?

Reply:
Ok so normalizing the predicates individually doesn't work, because sometimes we have two predicates like `<T as Foo>::Assoc: Bar` and `<T as Foo>::Assoc == SomeType` so that one predicate can only be normalized if the other predicate is in the param env. That's why the implementation of `param_env` does that whole normalizing the param env in itself thing. This also means that my implementation here will stop working once we stop eagerly normalizing the param env.

Also [one UI test](https://github.com/rust-lang/rust/blob/929bceeaf8d3252581499ac4bc9d964d8a1a6874/tests/ui/consts/issue-68264-overflow.rs) will overflow if we normalize the predicates individually.

Reply:
Alright I found a way that seems to work: We can just normalize the `predicates_of` individually inside the `param_env` of the item.